File: 01-ll-designs/message_queue/solutions/python/message_queue.py
"""
Simple In-Memory Message Queue Implementation in Python

This implementation provides a basic message queue system with support for:
- Multiple topics
- Multiple producers and consumers
- Thread-safe operations
- FIFO message ordering
- Subscription management
"""

import threading
import time
import uuid
from collections import defaultdict, deque
from typing import Dict, List, Callable, Optional, Any
from dataclasses import dataclass, field
from abc import ABC, abstractmethod


@dataclass
class Message:
    """Represents a message in the queue"""
    id: str = field(default_factory=lambda: str(uuid.uuid4()))
    topic: str = ""
    payload: str = ""
    timestamp: float = field(default_factory=time.time)
    headers: Dict[str, str] = field(default_factory=dict)


class MessageHandler(ABC):
    """Abstract base class for message handlers"""
    
    @abstractmethod
    def handle_message(self, message: Message) -> None:
        """Process a received message"""
        pass


class Consumer:
    """Message consumer that can subscribe to topics"""
    
    def __init__(self, consumer_id: str, handler: MessageHandler):
        self.id = consumer_id
        self.handler = handler
        self.subscribed_topics = set()
        self._active = True
    
    def on_message(self, message: Message) -> None:
        """Called when a message is received"""
        if self._active:
            try:
                self.handler.handle_message(message)
            except Exception as e:
                print(f"Error processing message {message.id}: {e}")
    
    def stop(self):
        """Stop the consumer"""
        self._active = False
    
    def is_active(self) -> bool:
        """Check if consumer is active"""
        return self._active


class Topic:
    """Represents a topic with its subscribers and message queue"""
    
    def __init__(self, name: str, max_size: int = 1000):
        self.name = name
        self.max_size = max_size
        self.messages = deque(maxlen=max_size)
        self.subscribers: List[Consumer] = []
        self._lock = threading.RLock()
        self._message_count = 0
    
    def add_message(self, message: Message) -> None:
        """Add a message to the topic and deliver to subscribers"""
        with self._lock:
            self.messages.append(message)
            self._message_count += 1
            self._deliver_message(message)
    
    def subscribe(self, consumer: Consumer) -> None:
        """Add a subscriber to this topic"""
        with self._lock:
            if consumer not in self.subscribers:
                self.subscribers.append(consumer)
                consumer.subscribed_topics.add(self.name)
    
    def unsubscribe(self, consumer: Consumer) -> None:
        """Remove a subscriber from this topic"""
        with self._lock:
            if consumer in self.subscribers:
                self.subscribers.remove(consumer)
                consumer.subscribed_topics.discard(self.name)
    
    def _deliver_message(self, message: Message) -> None:
        """Deliver message to all active subscribers"""
        # Create a copy of subscribers to avoid modification during iteration
        current_subscribers = self.subscribers.copy()
        
        for subscriber in current_subscribers:
            if subscriber.is_active():
                try:
                    # Deliver message in a separate thread to avoid blocking
                    threading.Thread(
                        target=subscriber.on_message,
                        args=(message,),
                        daemon=True
                    ).start()
                except Exception as e:
                    print(f"Error delivering message to {subscriber.id}: {e}")
            else:
                # Remove inactive subscribers
                self.unsubscribe(subscriber)
    
    def get_stats(self) -> Dict[str, Any]:
        """Get topic statistics"""
        with self._lock:
            return {
                'name': self.name,
                'message_count': self._message_count,
                'queue_size': len(self.messages),
                'subscriber_count': len(self.subscribers),
                'max_size': self.max_size
            }


class MessageQueue:
    """Main message queue broker"""
    
    def __init__(self):
        self.topics: Dict[str, Topic] = {}
        self.producers: List['Producer'] = []
        self.consumers: List[Consumer] = []
        self._lock = threading.RLock()
    
    def create_topic(self, name: str, max_size: int = 1000) -> Topic:
        """Create a new topic"""
        with self._lock:
            if name not in self.topics:
                self.topics[name] = Topic(name, max_size)
            return self.topics[name]
    
    def delete_topic(self, name: str) -> bool:
        """Delete a topic"""
        with self._lock:
            if name in self.topics:
                topic = self.topics[name]
                # Unsubscribe all consumers
                for consumer in topic.subscribers.copy():
                    topic.unsubscribe(consumer)
                del self.topics[name]
                return True
            return False
    
    def publish(self, topic_name: str, payload: str, headers: Optional[Dict[str, str]] = None) -> str:
        """Publish a message to a topic"""
        with self._lock:
            # Create topic if it doesn't exist
            if topic_name not in self.topics:
                self.create_topic(topic_name)
            
            message = Message(
                topic=topic_name,
                payload=payload,
                headers=headers or {}
            )
            
            self.topics[topic_name].add_message(message)
            return message.id
    
    def subscribe(self, consumer: Consumer, topic_name: str) -> None:
        """Subscribe a consumer to a topic"""
        with self._lock:
            # Create topic if it doesn't exist
            if topic_name not in self.topics:
                self.create_topic(topic_name)
            
            self.topics[topic_name].subscribe(consumer)
            
            # Add consumer to our list if not already present
            if consumer not in self.consumers:
                self.consumers.append(consumer)
    
    def unsubscribe(self, consumer: Consumer, topic_name: str) -> None:
        """Unsubscribe a consumer from a topic"""
        with self._lock:
            if topic_name in self.topics:
                self.topics[topic_name].unsubscribe(consumer)
    
    def get_topic_stats(self, topic_name: str) -> Optional[Dict[str, Any]]:
        """Get statistics for a specific topic"""
        with self._lock:
            if topic_name in self.topics:
                return self.topics[topic_name].get_stats()
            return None
    
    def get_all_stats(self) -> Dict[str, Any]:
        """Get statistics for all topics"""
        with self._lock:
            return {
                'topics': {name: topic.get_stats() for name, topic in self.topics.items()},
                'total_topics': len(self.topics),
                'total_consumers': len(self.consumers),
                'total_producers': len(self.producers)
            }


class Producer:
    """Message producer"""
    
    def __init__(self, producer_id: str, message_queue: MessageQueue):
        self.id = producer_id
        self.message_queue = message_queue
        # Register with message queue
        message_queue.producers.append(self)
    
    def publish(self, topic: str, payload: str, headers: Optional[Dict[str, str]] = None) -> str:
        """Publish a message to a topic"""
        return self.message_queue.publish(topic, payload, headers)


# Example usage and demonstration
class PrintMessageHandler(MessageHandler):
    """Simple message handler that prints messages"""
    
    def __init__(self, consumer_id: str):
        self.consumer_id = consumer_id
    
    def handle_message(self, message: Message) -> None:
        print(f"[{self.consumer_id}] Received message {message.id[:8]} on topic '{message.topic}': {message.payload}")


def demo():
    """Demonstrate the message queue system"""
    print("=== Message Queue Demo ===\n")
    
    # Create message queue
    mq = MessageQueue()
    
    # Create consumers
    consumer1 = Consumer("consumer-1", PrintMessageHandler("consumer-1"))
    consumer2 = Consumer("consumer-2", PrintMessageHandler("consumer-2"))
    consumer3 = Consumer("consumer-3", PrintMessageHandler("consumer-3"))
    
    # Create producer
    producer = Producer("producer-1", mq)
    
    # Subscribe consumers to topics
    print("Setting up subscriptions...")
    mq.subscribe(consumer1, "orders")
    mq.subscribe(consumer2, "orders")
    mq.subscribe(consumer3, "notifications")
    
    # Publish some messages
    print("\nPublishing messages...")
    producer.publish("orders", "Order #1001 created")
    producer.publish("orders", "Order #1002 created")
    producer.publish("notifications", "System maintenance scheduled")
    producer.publish("orders", "Order #1003 created")
    
    # Wait for message processing
    time.sleep(1)
    
    # Show statistics
    print("\n=== Statistics ===")
    stats = mq.get_all_stats()
    for topic_name, topic_stats in stats['topics'].items():
        print(f"Topic '{topic_name}': {topic_stats['message_count']} messages, {topic_stats['subscriber_count']} subscribers")
    
    # Unsubscribe a consumer
    print("\nUnsubscribing consumer-1 from orders...")
    mq.unsubscribe(consumer1, "orders")
    
    # Publish more messages
    print("Publishing more messages...")
    producer.publish("orders", "Order #1004 created")
    
    time.sleep(1)
    print("\nDemo completed!")


if __name__ == "__main__":
    demo()

File: 01-ll-designs/message_queue/solutions/go/message_queue.go
/*
Simple In-Memory Message Queue Implementation in Go

This implementation provides a thread-safe message queue system with support for:
- Multiple topics with concurrent access using channels and goroutines
- Producer-consumer patterns with multiple subscribers
- FIFO message ordering within topics
- Subscription management with dynamic subscribe/unsubscribe
- Statistics and monitoring capabilities
- Graceful shutdown and cleanup
*/

package main

import (
	"context"
	"fmt"
	"log"
	"math/rand"
	"sync"
	"sync/atomic"
	"time"
)

// generateID creates a simple random ID
func generateID() string {
	const charset = "abcdefghijklmnopqrstuvwxyz0123456789"
	b := make([]byte, 8)
	for i := range b {
		b[i] = charset[rand.Intn(len(charset))]
	}
	return string(b)
}

// Message represents a message in the queue
type Message struct {
	ID        string            `json:"id"`
	Topic     string            `json:"topic"`
	Payload   string            `json:"payload"`
	Timestamp time.Time         `json:"timestamp"`
	Headers   map[string]string `json:"headers"`
}

// NewMessage creates a new message
func NewMessage(topic, payload string, headers map[string]string) *Message {
	if headers == nil {
		headers = make(map[string]string)
	}
	return &Message{
		ID:        generateID(),
		Topic:     topic,
		Payload:   payload,
		Timestamp: time.Now(),
		Headers:   headers,
	}
}

// String returns a string representation of the message
func (m *Message) String() string {
	return fmt.Sprintf("Message{id='%s', topic='%s', payload='%s'}", 
		m.ID[:8], m.Topic, m.Payload)
}

// MessageHandler defines the interface for handling messages
type MessageHandler interface {
	HandleMessage(message *Message) error
}

// MessageHandlerFunc is a function adapter for MessageHandler
type MessageHandlerFunc func(message *Message) error

// HandleMessage implements MessageHandler
func (f MessageHandlerFunc) HandleMessage(message *Message) error {
	return f(message)
}

// Consumer represents a message consumer
type Consumer struct {
	id               string
	handler          MessageHandler
	subscribedTopics map[string]bool
	active           int32 // atomic boolean
	mu               sync.RWMutex
}

// NewConsumer creates a new consumer
func NewConsumer(id string, handler MessageHandler) *Consumer {
	return &Consumer{
		id:               id,
		handler:          handler,
		subscribedTopics: make(map[string]bool),
		active:           1,
	}
}

// OnMessage processes a received message
func (c *Consumer) OnMessage(message *Message) {
	if !c.IsActive() {
		return
	}
	
	go func() {
		defer func() {
			if r := recover(); r != nil {
				log.Printf("Panic in consumer %s processing message %s: %v", 
					c.id, message.ID, r)
			}
		}()
		
		if err := c.handler.HandleMessage(message); err != nil {
			log.Printf("Error in consumer %s processing message %s: %v", 
				c.id, message.ID, err)
		}
	}()
}

// Stop stops the consumer
func (c *Consumer) Stop() {
	atomic.StoreInt32(&c.active, 0)
}

// IsActive returns whether the consumer is active
func (c *Consumer) IsActive() bool {
	return atomic.LoadInt32(&c.active) == 1
}

// ID returns the consumer ID
func (c *Consumer) ID() string {
	return c.id
}

// GetSubscribedTopics returns a copy of subscribed topics
func (c *Consumer) GetSubscribedTopics() []string {
	c.mu.RLock()
	defer c.mu.RUnlock()
	
	topics := make([]string, 0, len(c.subscribedTopics))
	for topic := range c.subscribedTopics {
		topics = append(topics, topic)
	}
	return topics
}

// addSubscription adds a topic subscription (internal use)
func (c *Consumer) addSubscription(topic string) {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.subscribedTopics[topic] = true
}

// removeSubscription removes a topic subscription (internal use)
func (c *Consumer) removeSubscription(topic string) {
	c.mu.Lock()
	defer c.mu.Unlock()
	delete(c.subscribedTopics, topic)
}

// TopicStats represents statistics for a topic
type TopicStats struct {
	Name            string `json:"name"`
	MessageCount    int64  `json:"messageCount"`
	QueueSize       int    `json:"queueSize"`
	SubscriberCount int    `json:"subscriberCount"`
	MaxSize         int    `json:"maxSize"`
}

// Topic represents a message topic
type Topic struct {
	name         string
	maxSize      int
	messages     chan *Message
	subscribers  []*Consumer
	messageCount int64
	mu           sync.RWMutex
	ctx          context.Context
	cancel       context.CancelFunc
}

// NewTopic creates a new topic
func NewTopic(name string, maxSize int) *Topic {
	ctx, cancel := context.WithCancel(context.Background())
	return &Topic{
		name:        name,
		maxSize:     maxSize,
		messages:    make(chan *Message, maxSize),
		subscribers: make([]*Consumer, 0),
		ctx:         ctx,
		cancel:      cancel,
	}
}

// AddMessage adds a message to the topic
func (t *Topic) AddMessage(message *Message) bool {
	select {
	case t.messages <- message:
		atomic.AddInt64(&t.messageCount, 1)
		t.deliverMessage(message)
		return true
	default:
		log.Printf("Topic %s is full, dropping message: %s", t.name, message.ID)
		return false
	}
}

// Subscribe adds a consumer to the topic
func (t *Topic) Subscribe(consumer *Consumer) {
	t.mu.Lock()
	defer t.mu.Unlock()
	
	// Check if already subscribed
	for _, sub := range t.subscribers {
		if sub.ID() == consumer.ID() {
			return
		}
	}
	
	t.subscribers = append(t.subscribers, consumer)
	consumer.addSubscription(t.name)
}

// Unsubscribe removes a consumer from the topic
func (t *Topic) Unsubscribe(consumer *Consumer) {
	t.mu.Lock()
	defer t.mu.Unlock()
	
	for i, sub := range t.subscribers {
		if sub.ID() == consumer.ID() {
			// Remove from slice
			t.subscribers = append(t.subscribers[:i], t.subscribers[i+1:]...)
			consumer.removeSubscription(t.name)
			break
		}
	}
}

// deliverMessage delivers a message to all active subscribers
func (t *Topic) deliverMessage(message *Message) {
	t.mu.RLock()
	currentSubscribers := make([]*Consumer, len(t.subscribers))
	copy(currentSubscribers, t.subscribers)
	t.mu.RUnlock()
	
	for _, subscriber := range currentSubscribers {
		if subscriber.IsActive() {
			subscriber.OnMessage(message)
		} else {
			// Remove inactive subscribers
			t.Unsubscribe(subscriber)
		}
	}
}

// GetStats returns topic statistics
func (t *Topic) GetStats() TopicStats {
	t.mu.RLock()
	defer t.mu.RUnlock()
	
	return TopicStats{
		Name:            t.name,
		MessageCount:    atomic.LoadInt64(&t.messageCount),
		QueueSize:       len(t.messages),
		SubscriberCount: len(t.subscribers),
		MaxSize:         t.maxSize,
	}
}

// Close closes the topic and cleans up resources
func (t *Topic) Close() {
	t.cancel()
	close(t.messages)
}

// MessageQueue represents the main message queue broker
type MessageQueue struct {
	topics    map[string]*Topic
	consumers []*Consumer
	mu        sync.RWMutex
}

// NewMessageQueue creates a new message queue
func NewMessageQueue() *MessageQueue {
	return &MessageQueue{
		topics:    make(map[string]*Topic),
		consumers: make([]*Consumer, 0),
	}
}

// CreateTopic creates a new topic
func (mq *MessageQueue) CreateTopic(name string, maxSize int) *Topic {
	mq.mu.Lock()
	defer mq.mu.Unlock()
	
	if topic, exists := mq.topics[name]; exists {
		return topic
	}
	
	topic := NewTopic(name, maxSize)
	mq.topics[name] = topic
	return topic
}

// DeleteTopic deletes a topic
func (mq *MessageQueue) DeleteTopic(name string) bool {
	mq.mu.Lock()
	defer mq.mu.Unlock()
	
	topic, exists := mq.topics[name]
	if !exists {
		return false
	}
	
	// Unsubscribe all consumers
	for _, consumer := range mq.consumers {
		topic.Unsubscribe(consumer)
	}
	
	topic.Close()
	delete(mq.topics, name)
	return true
}

// Publish publishes a message to a topic
func (mq *MessageQueue) Publish(topicName, payload string, headers map[string]string) string {
	// Create topic if it doesn't exist
	topic := mq.CreateTopic(topicName, 1000)
	
	message := NewMessage(topicName, payload, headers)
	topic.AddMessage(message)
	return message.ID
}

// Subscribe subscribes a consumer to a topic
func (mq *MessageQueue) Subscribe(consumer *Consumer, topicName string) {
	// Create topic if it doesn't exist
	topic := mq.CreateTopic(topicName, 1000)
	topic.Subscribe(consumer)
	
	mq.mu.Lock()
	defer mq.mu.Unlock()
	
	// Add consumer to our list if not already present
	for _, c := range mq.consumers {
		if c.ID() == consumer.ID() {
			return
		}
	}
	mq.consumers = append(mq.consumers, consumer)
}

// Unsubscribe unsubscribes a consumer from a topic
func (mq *MessageQueue) Unsubscribe(consumer *Consumer, topicName string) {
	mq.mu.RLock()
	topic, exists := mq.topics[topicName]
	mq.mu.RUnlock()
	
	if exists {
		topic.Unsubscribe(consumer)
	}
}

// GetTopicStats returns statistics for a specific topic
func (mq *MessageQueue) GetTopicStats(topicName string) *TopicStats {
	mq.mu.RLock()
	topic, exists := mq.topics[topicName]
	mq.mu.RUnlock()
	
	if !exists {
		return nil
	}
	
	stats := topic.GetStats()
	return &stats
}

// GetAllStats returns statistics for all topics
func (mq *MessageQueue) GetAllStats() map[string]interface{} {
	mq.mu.RLock()
	defer mq.mu.RUnlock()
	
	topicStats := make(map[string]TopicStats)
	for name, topic := range mq.topics {
		topicStats[name] = topic.GetStats()
	}
	
	return map[string]interface{}{
		"topics":         topicStats,
		"totalTopics":    len(mq.topics),
		"totalConsumers": len(mq.consumers),
	}
}

// Close closes the message queue and all topics
func (mq *MessageQueue) Close() {
	mq.mu.Lock()
	defer mq.mu.Unlock()
	
	for _, topic := range mq.topics {
		topic.Close()
	}
	
	for _, consumer := range mq.consumers {
		consumer.Stop()
	}
}

// Producer represents a message producer
type Producer struct {
	id           string
	messageQueue *MessageQueue
}

// NewProducer creates a new producer
func NewProducer(id string, messageQueue *MessageQueue) *Producer {
	return &Producer{
		id:           id,
		messageQueue: messageQueue,
	}
}

// Publish publishes a message to a topic
func (p *Producer) Publish(topic, payload string, headers map[string]string) string {
	return p.messageQueue.Publish(topic, payload, headers)
}

// ID returns the producer ID
func (p *Producer) ID() string {
	return p.id
}

// PrintMessageHandler is a simple message handler for demonstration
type PrintMessageHandler struct {
	consumerID string
}

// NewPrintMessageHandler creates a new print message handler
func NewPrintMessageHandler(consumerID string) *PrintMessageHandler {
	return &PrintMessageHandler{consumerID: consumerID}
}

// HandleMessage implements MessageHandler
func (h *PrintMessageHandler) HandleMessage(message *Message) error {
	fmt.Printf("[%s] Received message %s on topic '%s': %s\n",
		h.consumerID, message.ID[:8], message.Topic, message.Payload)
	return nil
}

// Demo demonstrates the message queue system
func demo() {
	fmt.Println("=== Message Queue Demo ===\n")
	
	// Create message queue
	mq := NewMessageQueue()
	defer mq.Close()
	
	// Create consumers
	consumer1 := NewConsumer("consumer-1", NewPrintMessageHandler("consumer-1"))
	consumer2 := NewConsumer("consumer-2", NewPrintMessageHandler("consumer-2"))
	consumer3 := NewConsumer("consumer-3", NewPrintMessageHandler("consumer-3"))
	
	// Create producer
	producer := NewProducer("producer-1", mq)
	
	// Subscribe consumers to topics
	fmt.Println("Setting up subscriptions...")
	mq.Subscribe(consumer1, "orders")
	mq.Subscribe(consumer2, "orders")
	mq.Subscribe(consumer3, "notifications")
	
	// Publish some messages
	fmt.Println("\nPublishing messages...")
	producer.Publish("orders", "Order #1001 created", nil)
	producer.Publish("orders", "Order #1002 created", nil)
	producer.Publish("notifications", "System maintenance scheduled", nil)
	producer.Publish("orders", "Order #1003 created", nil)
	
	// Wait for message processing
	time.Sleep(1 * time.Second)
	
	// Show statistics
	fmt.Println("\n=== Statistics ===")
	stats := mq.GetAllStats()
	if topicStats, ok := stats["topics"].(map[string]TopicStats); ok {
		for topicName, topicStat := range topicStats {
			fmt.Printf("Topic '%s': %d messages, %d subscribers\n",
				topicName, topicStat.MessageCount, topicStat.SubscriberCount)
		}
	}
	
	// Unsubscribe a consumer
	fmt.Println("\nUnsubscribing consumer-1 from orders...")
	mq.Unsubscribe(consumer1, "orders")
	
	// Publish more messages
	fmt.Println("Publishing more messages...")
	producer.Publish("orders", "Order #1004 created", nil)
	
	time.Sleep(1 * time.Second)
	fmt.Println("\nDemo completed!")
}

func main() {
	rand.Seed(time.Now().UnixNano())
	demo()
}